### PR TITLE
Updated spark example with docker images moved to gcr.io/google_containers

### PR DESCRIPTION
**Pull request by mwielgus**
- Take Spark from the official Apache mirror instead of http://ci.radanalytics.io
- Upgrade Spark from 1.2.1 to 1.4.0
- Use kubectl logs instead of links dump to check master status
- Kubectl outputs in the new format

### CODE REVIEW

**Comment by roberthbailey** [comment]
@ihmccreery Can you walk through this on GCE & GKE and make sure it works? Thanks.

**Comment by mwielgus** [comment]
@ihmccreery, the docker images have NOT been pushed to gcr.io/google_containers yet so you may have to hack the PR a bit.

**Comment by ihmccreery** [comment]
@mwielgus Can you please push the Docker images?  Once they're live, I can run through this.

**Comment by mwielgus** [comment]
@ihmccreery I pushed the images and they seem to be OK.

File: examples/spark/README.md [hunk 1]
```diff
@@ -44,38 +42,44 @@ the Master pod.
 $ kubectl create -f examples/spark/spark-master-service.json
 ```
 
-Ensure that the Master service is running and functional.
+Ensure that the Master is running and functional.
```
Comment (ihmccreery) [inline examples/spark/README.md]: Please explain.  How?

Comment (ihmccreery) [inline examples/spark/README.md]: (Or just remove, because the next section takes care of this.)

File: examples/spark/README.md [hunk 2]
```diff
@@ -94,70 +98,53 @@ $ kubectl create -f examples/spark/spark-worker-controller.json
...
+```shell
+$ kubectl get service spark-master
+NAME           LABELS              SELECTOR            IP(S)          PORT(S)
+spark-master   name=spark-master   name=spark-master   10.0.204.187   7077/TCP
+```
 
-sh-4.2# export SPARK_LOCAL_HOSTNAME=$(hostname -i)
+SSH to one of your cluster nodes and run spark-base image. Inside of the image there is a script
+that sets up the environment besed on the provided ip and port of the Master. After running it
+you are good to go.
```
Comment (ihmccreery) [inline examples/spark/README.md]: There should be a better way to do this, (i.e. run straight from my machine, not ssh into a node).  Can we instead give them a client file to download, so they can connect straight from their machine?

File: examples/spark/README.md [hunk 3]
```diff
@@ -94,70 +98,53 @@ $ kubectl create -f examples/spark/spark-worker-controller.json
...
 
-sh-4.2# export SPARK_LOCAL_HOSTNAME=$(hostname -i)
+SSH to one of your cluster nodes and run spark-base image. Inside of the image there is a script
+that sets up the environment besed on the provided ip and port of the Master. After running it
+you are good to go. 
 
-sh-4.2# MASTER=spark://spark-master:7077 pyspark
-Python 2.7.5 (default, Jun 17 2014, 18:11:42)
-[GCC 4.8.2 20140120 (Red Hat 4.8.2-16)] on linux2
+```
```
Comment (ihmccreery) [inline examples/spark/README.md]: At the very least, we should explain how to `ssh` into a node.

File: examples/spark/README.md (hunk 2 above)
Comment (ihmccreery) [inline examples/spark/README.md]: Presumably you'll have to open a LoadBalancer in the YAML files.  See [the Guestbook Example](https://github.com/GoogleCloudPlatform/kubernetes/tree/master/examples/guestbook#using-type-loadbalancer-for-the-frontend-service-cloud-provider-specific).

**Comment by ihmccreery** [comment]
@mwielgus See comments above.  Otherwise, LGTM.  Checked on GKE; I'll wait to re-check on GCE when the comments are resolved.

File: examples/spark/README.md (hunk 2 above)
Comment (mwielgus) [inline examples/spark/README.md]: The extra steps for installing java, scala, spark etc and configuring everything on the the user's machine will make everything much more complicated.

File: examples/spark/README.md (hunk 3 above)
Comment (mwielgus) [inline examples/spark/README.md]: Done.

File: examples/spark/README.md (hunk 2 above)
Comment (ihmccreery) [inline examples/spark/README.md]: Couldn't we ask them to install Docker, (which they'll need to do much with K8s anyway,) and run the same image you pull here?

Comment (mwielgus) [inline examples/spark/README.md]: Yes, we could add links to installation guides on Mac, Windows, Linux (and have them download 0.5 gb of stuff from the internet) but would it really make the example easier?

Comment (ihmccreery) [inline examples/spark/README.md]: I think it would be more in line with how Kubernetes should be used—`ssh`-ing into a node doesn't seem like a workflow we should encourage our users to do.  I'd recommend putting it in prerequisites; "You need [an installation of Docker](https://docs.docker.com/installation/) to run this example."  Bump to @RichieEscarez for a third set of eyes.

**Comment by RichieEscarez** [comment]
A couple things (including list of fixes below). 

We need to figure out whats going to be most beneficial for the user as far as what they can learn/take from this example and then use (the ease of the example is not as important). Im not familiar with Apache Spark at all, so my question is: "What's the expected/common use scenario for installing the Spark base image?" 

If we expect users to want to install the base image on a node on their local machine, then we should focus this example in that direction. Otherwise, if the common use scenario is that user will SSH into a remote node in order to install the base image, than we should build that up more. So first, we need to determine what the common use scenario is, and then help them learn how its done. The last option would be to address both scenarios add in a short alternative option (see example below). 

Either way, we should improve "Step Three" to make it clear what we are going to accomplish and why.

**Example** (remember i dont know spark so this might be technically incorrect):
--------"example-start"-------------

**Step Three: Install the Spark base image on a worker node**

Depending on the getting started guide that you used to create your cluster, you can install the Spark base image on a remote node or local node.

 - If you are using a cloud provider like Google Compute Engine (GCE) or Google Container Engine (GKE), then you can SSH into a remote node and install the Spark base image.
 
        show-SSH-steps-here

 - If you have a cluster running locally, you can download and install Docker[URL here to download docker] on your local machine before installing the Spark base image on a node.

        define-localinstall-steps-here

--------"example-end"---------

**Other fixes we should make to improve this doc:**

- In Step One, Add a link to the file "examples/spark/spark-master-service.json" (like the link to the ...spark-master.json file)
- In Step Two, explain more about why creating a rep controller results in new worker pods (ie. *the rc keeps your pods alive and automatically creates the worker pods when it finds that the number of pods does not equal the number of replicas specified in the pod spec.*)
- In Step Three:
    - Provide (1) the "kubectl get nodes" code example and output, plus (2) glcoud compute ssh example with name from "get nodes" ouput:  

    "For example, run: 
	
	    kubectl get nodes
	    show-output-here
	
		gcloud comput ssh NAMEFROMOUTPUT
		show-ouput-here

    - Fix typo: "besed" -> "based"
    - use capital "IP"
    - The TLDR section does not point them back to installing the spark base image
    - Replace "After running it you are good to go." With a paragraph at the very end of the example (maybe a separate section "Result"). Something like:  
    "Result:
	You now have services, replication controllers, and pods for the Spark master and Spark workers. You can take this example to the next step and start using the Apache Spark cluster you just created, see [point to spark docs here] for more information."

**Comment by mwielgus** [comment]
@RichieEscarez Thank you for the review.

The most convenient (from the developer's perspective) would be to have Spark client running on the desktop. But it would require much more steps to have it up and running and would be rather focused no things unrelated to Kubernetes itself like docker installation on mac/windows/linux, setting up Spark's security so that not everyone has the access to the spark cluster. I think we should not go this way.

So the other option - sshing to the cluster - is IMHO much easier. I added the steps describing how to ssh to a node on GCE (and applied you other comments). But ssh should be rather covered in details on a separate doc page, so that the other examples that also involve sshing to the cluster can point to it too. And speaking about examples in general - I think this one is quite decent now and I would rather focus on fixing others before the launch instead of adding all possible use scenarios to this particular one.

**Comment by RichieEscarez** [comment]
@mwielgus, I agree on all points.

- This example should remain simple (perhaps in the future, maybe an advanced version can be created to build on and cover the more complicated scenario?).
- We should eventually create a standalone "how to SSH into a node" document that others can point to/reuse (ive found that a few of other examples also SSH into nodes). *- ill open an issue for about this*
- From my point of view LGTM (pending: @ihmccreery to check on GCE). Ultimately, this doc has been improved/fixed and we should move on to the next "to do" (although we can spend more time editing it [*for consistency with other examples, ie formatting, structure, etc.*] but lets save that for post v1 when there are fewer competing priorities).

**Comment by RichieEscarez** [comment]
@ihmccreery, can you give the "ok-to-merge" when this is all tested and ready to go?

**Comment by ihmccreery** [comment]
LGTM, checked on GCE & GKE.  We don't need **ok-to-merge** because this is only a docs change.  Thanks to you both for contributing so much to this doc.

**Comment by yujuhong** [comment]
> LGTM, checked on GCE & GKE. We don't need ok-to-merge because this is only a docs change. Thanks to you both for contributing so much to this doc.

This PR seems to be more than a docs change as it touches scripts too :) 
@ihmccreery (or @roberthbailey), please label it as "ok-to-merge" when the PR is ready. Thanks!

@mwielgus, shippable complains about the boilerplate header. Please resolve this, thanks.
```
 ./hack/verify-boilerplate.sh
Boilerplate header is wrong for: ./examples/spark/images/base/setup_client.sh
Boilerplate header is wrong for: ./examples/spark/images/worker/start.sh
Boilerplate header is wrong for: ./examples/spark/images/master/start.sh
```

**Comment by mwielgus** [comment]
Boilerplate fixed.

**Comment by ihmccreery** [comment]
I can't give **ok-to-merge** because I'm not a tech lead.  Assigning to @brendandburns.

**Comment by brendandburns** [comment]
Ok to merge.
